### ISSUE TITLE
SAM start/stop

### ISSUE THREAD

**Issue by clburlison** [versx/iPhoneController#7]
As part of #4 @kamieniarz thinks it would be cool for if you supported starting/stopping SAM. If you install Apple Configurator there is a separate ID that is used to target specific devices. This isn't the uuid used with other tools so you'll have to parse the output from cfgutil list(?)

Can upload the profile when not on mobile or you can search for the following text in one of our mutual servers. 

```
cfgutil -v remove-profile com.apple.configurator.singleappmode
cfgutil -v install-profile ~/Desktop/sam_pogo.mobileconfig
```

> **This requires Apple Configurator 2 to be installed**
If anyone wants to wrap that into their internal tooling. Might make setup for new iphones easier. Could also be used as a potential fix for the "Guided access not available" popup shit that doesn't require a reboot.
> 
> __**Provided as is. I'm not support. Just sharing to help the community**__
> 
> Install results in exit code 1 and a dumb message about touching the device but it still works. Apple bugs 🤷‍♂️
> 
> Still doesn't help with system pop ups but those are mostly triggered due to a reboot over here 🤷‍♂️

**Comment by clburlison** [comment]
Attaching a config profile for starting SAM on an iPhone. This is a newer version that includes the exact settings that Apple Configurator 2 applies when doing the operation via the application. I pulled that from an iOS backup. The remove command can use the profile id so always use the below domain. The annoying part will be calling the `list` command in order to find the ecid values.  

```
# Get ecids
cfgutil --format JSON list
{"Command":"list","Output":{"0xXXXXXXXXXXX":{"locationID":12341234,"UDID":"00000000-0000000000000000","ECID":"0x0000000000000","name":"swift-sucks","deviceType":"iPhone11,6"}},"Type":"CommandOutput","Devices":["0xXXXXXXXXXXX"]}

# Remove/install profiles
cfgutil -e 0x0000000000000 -e 0x0000000000001 remove-profile com.apple.configurator.singleappmode
cfgutil -e 0x0000000000000 -e 0x0000000000001 install-profile ~/Desktop/sam_pogo.mobileconfig
```

Last note it might be useful to include option -C and -K commands to pass the supervision org data. For me this was never required since it's stored in my keychain however some users needed it? 

```
Global options:
	[-C | --certificate] <argument>
	Path to DER-encoded certificate of supervising organization.

	[-K | --private-key] <argument>
	Path to DER-encoded private key of supervising organization.

IE - 
Apple Config > Prefs > Orgs > click on Org > Export Supervision Identity > format > Unencrypted DER
cfgutil -e 0x0000000000000 -e 0x0000000000001 -C ~/Desktop/my_org.crt -K ~/Desktop/my_org.der install-profile ~/Desktop/sam_pogo.mobileconfig
```

Also, feel free to close this issue if you don't want to work on it. This has no concern to me just trying to make the information available in an easy format. 

[Configurator Single App Mode: Pokémon GO.mobileconfig.zip](https://github.com/versx/iPhoneController/files/5600523/Configurator.Single.App.Mode.Pokemon.GO.mobileconfig.zip)

**Comment by kamieniarz** [comment]
Thanks for bringing that up. I think this feature might help mappers a lot with theirs devices that have Touch ID issues. Restarting SAM when not at home (in my case via teamviewer atm) helps fixing 99% of all issues with these devices so it would be great to make it run with simple Discord command

**Comment by Kneckter** [comment]
Putting this here so I don't forget: When the game freezes, removing/adding the profile will not revive the game. Removing, apply a different one for like Calc, removing, and then applying the pogo one does revive the game. This can help save Touch ID devices that cannot be rebooted and have a frozen game. The issue is that there are twice as many profile operations.

**Comment by TRGoCPftF** [comment]
> Configurator Single App Mode: Pokémon GO.mobileconfig.zip

How did you go about adding the AppLock configuration to the profile? Did you manually add the [AppLock keys](https://developer.apple.com/documentation/devicemanagement/applock/app?language=objc) to the profile?

You could painlessly build a map of Device Name, UDID, and ECID with 
`cfgutil -f --format plist get name ECID UDID > AFile.plist` 
Using plist or raw text seems easier/cleaner as the JSON output is a dictionary of results for each device, but the key for each device is the ECID of the device, so not as easy to parse through as Text or plist out.

Also, considering the profile to enable SAM/AppLock is based on the BundleID, have you tested to see what happens if you push the profile when the application with the specified bundle is not available? I'm curious to see if it ends up in the black "Guided Access Mode Unavailable" screen or what the results are.

Another thing that'd be smart to avoid that/validate before moving is insuring the app is actually installed 
`cfgutil -e 0xAnECID get installedApps`
Will give you a dict with an array of every installed application like
`{"itunesName":"TikTok",
"displayName":"TikTok",
"bundleIdentifier":"com.zhiliaoapp.musically",
"bundleVersion":"183015"}`

**Comment by Kneckter** [comment]
In the DCMRemoteListener PR, I used the cfgutil to grab the first 3 variables instead of using ios-deploy. I also found the db that maps tethered IPs to devices so grabbing the IPs went a lot smoother. 

With the .zip profile here, I received an invalid profile warning so I made new ones. The app-lock does have a bundle ID but it has been working on multiple iOS versions across multiple device models. It doesn't like when you try to push the SAM profile over an existing one and if it isn't valid, it doesn't install.

You guys can check the files I used in the DCMRemoteListener PR to see how I made it work. https://github.com/versx/DCMRemoteListener/pull/3. Just a warning, I don't think it likes when you try to reapply 100 profiles at once.
